src/aws-route53/split-horizon-dns.ts

```diff
@@ -0,0 +1,144 @@
+
+export interface AliasTarget {
+  readonly target: route53.IAliasRecordTarget | Array<string>;
+  readonly private?: boolean;
+  readonly public?: boolean;
+  readonly ttl?: cdk.Duration;
+}
+
+/**
+ * @param zoneName The DNS name of the zone to create
+ * @param existingPublicZone An existing public zone to use instead of creating a new one
+ * @param existingPrivateZone An existing private zone to use instead of creating a new one
+ * @param disallowPrivateZone Override the default behavior of creating a private zone. Will also block adding private records.
+ * @param includeCertificate Whether to create an ACM certificate for the zone
+ * @param certAlternateNames Alternate names to include in the certificate
+ * @param privateZoneVpcs VPCs to associate with the private zone
+ * @param targets Targets to create A records for
+ */
+export interface ISplitHorizonDnsProps {
+  readonly zoneName: string;
+  readonly recordName?: string;
+  readonly existingPublicZone?: route53.IHostedZone;
+  readonly existingPrivateZone?: route53.IHostedZone;
+  readonly disallowPrivateZone?: boolean;
+  readonly certAlternateNames?: Array<string>;
+  readonly privateZoneVpcs?: Array<ec2.Vpc>;
+  readonly targets: Array<AliasTarget>;
+  readonly includeCertificate?: boolean;
+}
+
+export interface ISplitHorizonDns {
+  publicZone: route53.IHostedZone;
+  privateZone?: route53.IHostedZone;
+  records: Array<ARecordArray>;
+}
+
+/**
+ * Creates a public and private zone for a given domain name, and creates A records for the given targets.
+ * @property publicZone The public zone created
+ * @property privateZone The private zone created
+ * @property records The A records created
+ */
+export class SplitHorizonDns extends Construct implements ISplitHorizonDns {
+  public publicZone: route53.IHostedZone;
+
+  public privateZone?: route53.IHostedZone;
+
+  public records: Array<ARecordArray>;
+
+  public certificate?: acm.ICertificate;
+
+  constructor(scope: Construct, id: string, private props: ISplitHorizonDnsProps) {
+    super(scope, id);
+
+    const {
+      zoneName,
+      existingPublicZone,
+      existingPrivateZone,
+      disallowPrivateZone,
+      includeCertificate,
+      certAlternateNames,
+      privateZoneVpcs,
+      targets,
+    } = this.props;
+
+    if (existingPublicZone) {
+      this.publicZone = existingPublicZone;
+    } else {
+      this.publicZone = new route53.HostedZone(this, 'PublicZone', {
+        zoneName: zoneName,
+      });
+    }
+
+    if (includeCertificate) {
+      this.certificate = new acm.Certificate(this, 'Certificate', {
+        domainName: zoneName,
+        subjectAlternativeNames: certAlternateNames,
+      });
+    }
+
+    if (disallowPrivateZone) {
+      console.log('Private zone creation is disallowed. Skipping...');
+    } else if (existingPrivateZone) {
+      this.privateZone = existingPrivateZone;
+    } else {
+      this.privateZone = new route53.HostedZone(this, 'PrivateZone', {
```

Comment:
Please extract the creation of sub-constructs into a factory method that can be overridden in a subclass